### ISSUE TITLE
on_walk_over() for nodes (or better, on_new_standing_node() for players)

### ISSUE THREAD

**Issue by PilzAdam** [luanti-org/luanti#247]
There should be a function in node definition that is executed when the player walks over the node.

**Comment by paramat** [comment]
Good for creating trails of nodes with footsteps :+1:

**Comment by Ferk** [comment]
Would it be possible to replace the "damage_per_second" flag with a "on_collision_step" function?

**Comment by Calinou** [comment]
Is it possible to add this callback without increasing server resource usage too much? Since movement is client-side, a regular check needs to be done to know if a player (or another entity) is stepping on a node. Isn't this too resource-intensive?

**Comment by ObaniGemini** [comment]
https://github.com/minetest/minetest/issues/3918 is a better solution imo

**Comment by 0-afflatus** [comment]
Implementation aside, it's a great idea.

**Comment by C1ffisme** [comment]
Taking away my +1, I think #3918 is probably better. I'm not quite sure why I thumbed up this one.

**Comment by Ferk** [comment]
Better for what purpose?
For things like pressure plates it would mean we have to register a callback for the player's on_collide for every element that we want to check for... and every time the player collides with something all the checks would have to be performed separately in lua.

Also I got the impression #3918 is asking for a function as property in the definition of the entity, not an event that you can register multiple callbacks to.

**Comment by tenplus1** [comment]
wouldnt a global player table be easier that would hold node information on what player stands on, foot level and head level, then mods can access this table which is created in the engine itself for speed and detail what to do when say head is in water or standing on pressure plate ? we dont need callbacks for every single fuction.

**Comment by paramat** [comment]
Removing my +1 also, as this would seem to be intensive.

**Comment by Ferk** [comment]
@tenplus1 And when do you check that table? in the global_on_step? That's already a callback. You need to add callbacks somewhere for the mod to act, and the engine needs to retrieve the information every time to keep the table up to date.

Using #3918 to implement pressure plates might be less efficient. Because you would need to call the lua function in every collision.

But in the case of a specific on_walk_over registered for particular nodes, since the engine knows what specific nodes require it, it might be faster.
But I don't know enough about the engine design to be able to tell...

**Comment by paramat** [comment]
Good for footprints too.

**Comment by paramat** [comment]
We could use this position https://github.com/minetest/minetest/blob/master/src/localplayer.cpp#L674 and add a callback for the node that is 'got' for footstep souns.

**Comment by paramat** [comment]
I started an attempt but it will be more complex than i had expected, unsure if i will be able to code this.
It seems we need to store every player position and disable 'walked' callbacks if the player has not moved, just like a lua mod does (such as my trail mod).

**Comment by SmallJoker** [comment]
@paramat Why so complicated? Call `getStandingNodePos()` in `clientenvironment.cpp` after the collisions are handled and compare it with a previous position value. If there's a change, run the Lua callback.

**Comment by paramat** [comment]
Thanks.

**Comment by paramat** [comment]
@SmallJoker wouldn't that be clientside code though? This needs to be serverside.

**Comment by SmallJoker** [comment]
@paramat Yes, but if we tell the client that a node uses that field, a simple packet consisting the position could be sent each time the player enters/leaves that type of node, so the server does not have to rely on the movement updates.

**Comment by Ferk** [comment]
If something like #6219 was possible, the client position would be checked server side already. So I think there would be certain assurance that the client won't be able to trick the server into activating something it hasn't step over.

However, it might be able to trick the server into not activating something that it has indeed step over. But perhaps it's still a good enough compromise, specially if mods use this feature in such a way that activating the on_walk_over is the desirable option, and there was little incentive for cheaters to want to not activate it.

**Comment by paramat** [comment]
I think i prefer a completely serverside solution for obvioous reasons, no need for another packet too.

**Comment by tacotexmex** [comment]
Did it happen?

**Comment by DS-Minetest** [comment]
I don't think so.
But it is close to being done, it can only be a matter of a handful of years.

**Comment by paramat** [comment]
LOL silly me. No it's not done yet.

**Comment by paramat** [comment]
I was thinking about this feature today and surprised we still do not have it.
You may know i wrote a mod called 'trail' that added footprints to walked-on nodes and then eroded them to hardpacked dirt. I would like a higher performance engine API for this.

In this discussion it seems we prefer the implementation to be a function that is called when any player moves to a new 'standing node' (the node a player stands on), instead of the original suggestion that would require every node to constantly check for passing players above, which seems far more intensive, especially for mulitple nodes.

But, do we include lua entities too, such as mobs? I guess we should. Is the recently added 'on collide' feature #9668 for entities relevant or useful?

A completely serverside implementation seems attractive for simplicity, but does that mean a little more network delay? I doubt it makes much difference because:
If the 'standing node' checking code was clientside a packet would be sent to the server to run the callback, the result of walking on a node would be sent back to the client, so network delay would happen. I cannot see how this is any faster than just sending the existing movement packet to the server and the 'standing node' checking code being run in Server Active Object code.

But if serverside, the server has to deal with mulitple players instead of 1, so it is more difficult to check whether the 'standing node' has changed for a player, a table of positions for every player would be needed, or perhaps there is a better way to do this serverside? Maybe clientside is best after all?

**Comment by Bituvo** [comment]
Still waiting for this... 👀

**Comment by sfan5** [comment]
Low priority since mods can easily do this themselves.

**Comment by Bituvo** [comment]
There are many things that mods can do themselves that are also implemented into the Minetest API. For example, why set acceleration of an entity when you can add velocity each step? The only methods I've seen to accomplish node walk detection are clunky, to say the least.

**Comment by appgurueu** [comment]
> There are many things that mods can do themselves that are also implemented into the Minetest API. For example, why set acceleration of an entity when you can add velocity each step?

Because acceleration can then be client predicted, which is necessary for it to be smooth. Note also that just adding velocity each step is *NOT* correct because the object should also have accelerated *during* the step, so you also need to add $\frac{1}{2}at^2$ to the position. And if you did that it would be very janky using a typical ~100 ms serverstep.

This is very different from this feature request: There *can't* really be smooth client prediction with such a callback, because the server needs to tell the client what to do. So you incur at least one round trip. The *only* advantage in terms of jank could be clients self-reporting which nodes they walked over, but that is problematic when it comes to cheating, so you probably do want to do this serverside.

> The only methods I've seen to accomplish node walk detection are clunky, to say the least.

Since the addition of `get_node_boxes`, this should be fairly easy to implement properly. Note also that once any modder has implemented it properly, they (or someone else who finds it useful) can release it as (part of) a library mod.

**Comment by Bituvo** [comment]
I was not quite aware of that function, thanks for pointing that out! I guess this is pretty easily implemented then

**Comment by hlqkj** [comment]
It's true that this can be easily implemented by mods, and I've seen this done quite cleanly.

The best use case I've seen, was to change the player walking speed based on the ground type, much like the `move_resistance` field in a node definition table. This feature is, in my opinion, a *big* miss for Minetest, and while it's true that it can be implemented by mods, such implementations just can't be smooth enough, due to lack of client-side prediction.

I understand that this is totally off-topic here, but I thought to cite this here because I figure (probably being mistaken?) that having the internals related to this feature request implemented in the engine would also pave the way for such a movement speed modifier for nodes *below* the player.

**Comment by rubenwardy** [comment]
This issue is so old that I feel like if it's truly low priority it should just be closed
